### PR TITLE
Change the spec to use HTML's navigables instead of browsing contexts

### PR DESCRIPTION
**Pull request by OrKoN**
This PR updates the specification to define interactions with browsing contexts in terms of HTML spec's navigables without changing the API  (which keeps using "browsingContext" and "context" to refer to navigables).

Closes https://github.com/w3c/webdriver-bidi/issues/91

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/pull/729.html" title="Last updated on Jul 17, 2024, 10:25 AM UTC (a8145a6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/729/0fa2d99...a8145a6.html" title="Last updated on Jul 17, 2024, 10:25 AM UTC (a8145a6)">Diff</a>

### CODE REVIEW

**Comment by OrKoN** [comment]
TODOs:

1) check lifetime of the navigable in context destroyed
2) is a storage partition associated with the navigable?
3) check document's navigables references

**Review by jgraham — CHANGES_REQUESTED** [review]
Generally looking good, thanks! Just a few cases where we're linking to the wrong thing or inventing a concept that doesn't seem to exist.

**Comment by OrKoN** [comment]
@jgraham PTAL again

File: index.bs [hunk 1]
```diff
@@ -3091,13 +3067,13 @@ The [=remote end steps=] with |command parameters| are:
      field of |command parameters|, if present, or null otherwise.
 
   1. If |reference context id| is not null, let |reference context| be the
```
Comment (sadym-chromium) [inline index.bs:3069]: should `reference context id` be renamed as well?

File: index.bs [hunk 2]
```diff
@@ -1437,7 +1434,7 @@ To <dfn>cleanup the session</dfn> given |session|:
 
 <div algorithm>
 To <dfn>update the event map</dfn>, given
-|session|, |requested event names|, |browsing contexts|, and |enabled|:
+|session|, |requested event names|, |navigables|, and |enabled|:
 
 Note: The return value of this algorithm is a map between event names and
 contexts. When the events are being enabled, the contexts in the return value
```
Comment (sadym-chromium) [inline index.bs:1440]:
```suggestion
Note: The return value of this algorithm is a map between event names and
navigables. When the events are being enabled, the contexts in the return value
```

File: index.bs [hunk 3]
```diff
@@ -2893,13 +2869,13 @@ To <dfn>get the origin rectangle</dfn> given |document| and |origin|:
...
-1. Let |context id| be the value of the <code>context</code> field of
+1. Let |navigable id| be the value of the <code>context</code> field of
    |command parameters| if present, or null otherwise.
 
-1. If the implementation is unable to capture a screenshot of |context| for any
+1. If the implementation is unable to capture a screenshot of |navigable id| for any
    reason then return [=error=] with [=error code=] [=unsupported operation=].
 
-1. Let |document| be |context|'s [=active document=].
+1. Let |document| be |navigable id|'s [=active document=].
```
Comment (sadym-chromium) [inline index.bs:2878]: Missing getting navigable by it's ID

File: index.bs [hunk 4]
```diff
@@ -1491,16 +1488,16 @@ events are disabled, the return value is always empty.
...
-  1. For each |context id| in |browsing contexts|:
+  1. For each |navigable id| in |navigables|:
 
-    1. Let |context| be the result of [=trying=] to [=get a browsing context=]
-       with |context id|.
+    1. Let |navigable| be the result of [=trying=] to [=get a navigable=]
+       with |navigable id|.
 
-    1. Let |top-level context| be the [=top-level browsing context=] for |context|.
+    1. Let |top-level context| be the [=navigable/top-level traversable=] for |navigable|.
```
[the remark this answers is not in the document]
Comment (OrKoN) [reply index.bs:1500]: I think HTML does not define top-level navigable only traversables which are also navigables

File: index.bs [hunk 5]
```diff
@@ -3091,13 +3067,13 @@ The [=remote end steps=] with |command parameters| are:
      field of |command parameters|, if present, or null otherwise.
 
   1. If |reference context id| is not null, let |reference context| be the
-     result of [=trying=] to [=get a browsing context=] with
+     result of [=trying=] to [=get a navigable=] with
      |reference context id|. Otherwise let |reference context| be null.
```
Comment (sadym-chromium) [inline index.bs:3071]: `reference context` -> `reference navigable` ?

File: index.bs (hunk 3 above)
Comment (OrKoN) [reply index.bs:2878]: Looks like an existing bug in the spec?

File: index.bs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply index.bs:1500]: I meant the `|top-level context|` variable. But this can be out of scope of the PR

File: index.bs [hunk 6]
```diff
@@ -686,40 +681,40 @@ occurred on the [=remote end=].
...
 A [=BiDi session=] has a <dfn export for=event>global event set</dfn>
 which is a [=/set=] containing the event names for events that are enabled for all
-browsing contexts. This initially contains the [=event name=] for events that
+navigables. This initially contains the [=event name=] for events that
 are <dfn export for=event>in the default event set</dfn>.
 
-A [=BiDi session=] has a <dfn export for=event>browsing context event
-map</dfn>, which is a [=/map=] with [=/top-level browsing context=] keys and values
+A [=BiDi session=] has a <dfn export for=event>navigable event map</dfn>,
+which is a [=/map=] with [=navigable/top-level traversable=] keys and values
```
Comment (sadym-chromium) [inline index.bs:688]: It looks like the link is broken:
```suggestion
which is a [=/map=] with [=navigable/top-level-traversable=] keys and values
```

File: index.bs [hunk 7]
```diff
@@ -1061,21 +1054,25 @@ To <dfn>handle an incoming message</dfn> given a [=WebSocket connection=]
...
+1. Let |related navigables| be an empty [=/set=].
 
-1. If |settings|' [=relevant global object=] is a {{Window}}, append |settings|'
-   [=relevant global object=]'s <a>associated <code>Document</code></a>'s
-   [=Document/browsing context=] to |related browsing contexts|.
+1. Let |navigable| be null.
+
+1. If |settings|' [=relevant global object=] is a {{Window}}, set |navigable|
+   to [=relevant global object=]'s <a>associated <code>Document</code></a>'s
+   |=/node navigable=|.
```
Comment (sadym-chromium) [inline index.bs:1066]: Square brackets?
```suggestion
   [=/node navigable=].
```

File: index.bs (hunk 6 above)
Comment (OrKoN) [reply index.bs:688]: this line works for me and links to https://html.spec.whatwg.org/multipage/document-sequences.html#nav-top

Comment (sadym-chromium) [reply index.bs:692]: But it should be https://html.spec.whatwg.org/multipage/document-sequences.html#top-level-traversable

**Review by sadym-chromium — APPROVED** [review]

File: index.bs (hunk 6 above)
Comment (OrKoN) [reply index.bs]: Changed to `[=/top-level traversable=]` when references without a relation to a navigable

File: index.bs [hunk 8]
```diff
@@ -3154,21 +3134,21 @@ The [=remote end steps=] with |command parameters| are:
...
+  1. If |reference navigable id| is not null, let |reference navigable| be the
+     result of [=trying=] to [=get a navigable=] with
+     |reference navigable id|. Otherwise let |reference navigable| be null.
 
-  1. If |reference context| is not null and is not a [=top-level browsing context=],
+  1. If |reference navigable| is not null and is not a [=/top-level traversable=],
     return [=error=] with [=error code=] [=invalid argument=].
 
-  1. If the implementation is unable to create a new browsing context for any
+  1. If the implementation is unable to create a new navigable for any
```
Comment (jgraham) [inline index.bs:3147]:
```suggestion
  1. If the implementation is unable to create a new [=/top-level traversable=] for any
```

File: index.bs [hunk 9]
```diff
@@ -3184,42 +3164,38 @@ The [=remote end steps=] with |command parameters| are:
 
   <!-- This is based on step 5 of https://w3c.github.io/webdriver/#new-window,
        but without using the "current browsing context" concept. -->
-  1. Create a new [=top-level browsing context=] by running
-     the [=/create a new top-level traversable=] steps with null and empty string,
+  1. Let |navigable| be the result of trying to [=/create a new top-level traversable=] steps with null and empty string,
      and setting the [=associated user context=] for the newly created [=/top-level traversable=] to |user context|.
-     Which OS window the new [=/browsing context=] is created in depends on |type| and |reference context|:
+     Which OS window the new [=/navigable=] is created in depends on |type| and |reference navigable|:
```
Comment (jgraham) [inline index.bs:3169]: Should we use top-level traversable throughout this step and set of conditions?

File: index.bs [hunk 10]
```diff
@@ -4386,13 +4353,13 @@ navigated</dfn> steps given |context| and |navigation status|:
 
 1. Let |navigation id| be |navigation status|'s id.
 
-1. Let |related browsing contexts| be a [=/set=] containing |context|.
+1. Let |related navigable| be a [=/set=] containing navigable|.
```
Comment (jgraham) [inline index.bs:4356]:
```suggestion
1. Let |related navigable| be a [=/set=] containing |navigable|.
```

File: index.bs [hunk 11]
```diff
@@ -7902,26 +7867,28 @@ To <dfn>get the realm info</dfn> given |environment settings|:
     <dd>
       1. Let |document| be |environment settings|' [=relevant global object=]'s
          <a>associated <code>Document</code></a>.
-      1. Let |browsing context| be |document|'s [=/browsing context=].
-      1. Let |context id| be the [=browsing context id=] for |browsing context|.
+      1. Let |navigable| be |document|'s [=/node navigable=].
+      1. If |navigable| is null, return null.
```
Comment (jgraham) [inline index.bs:7871]: You need to update the callers to handle `null` as a return value.

**Review by jgraham — CHANGES_REQUESTED** [review]
Really close, just a few small issues that I noticed.

File: index.bs (hunk 11 above)
Comment (OrKoN) [reply index.bs:7871]: I believe most(some) of them do, did you spot one that is not handling it?

File: index.bs (hunk 9 above)
Comment (OrKoN) [reply index.bs]: Updated

File: index.bs (hunk 11 above)
Comment (OrKoN) [reply index.bs:7873]: Found missing places, and updated

**Review by jgraham — APPROVED** [review]
Thanks!

**Comment by jgraham** [comment]
Since this is basically a (large) editorial change, and the underlying problem is that rejoining the group is taking some time, I this this should be fine to merge.

**Pull request merged**